### PR TITLE
Make sure level_iters_has correct size before [] assignment

### PR DESCRIPTION
**Pull request by yuslepukhin**
@rven1 @siying @igorcanadi Attempt number two. My apologies for a wrong fix first. Please, review.

### CODE REVIEW

**Comment by yuslepukhin** [comment]
@rven1 Generally, the code might expect that that vector always has enough of the size according to the number of levels but apparently this is not the case and there might be another place there the bug is. On a side note operator[] is an unchecked access and it is usually used in a loop which is bounded by the size of that same vector but through out the code it is not the case.

File: db/forward_iterator.cc [hunk 1]
```diff
@@ -575,11 +575,17 @@ void ForwardIterator::BuildLevelIterators(const VersionStorageInfo* vstorage) {
          (user_comparator_->Compare(*read_options_.iterate_upper_bound,
                                     level_files[0]->smallest.user_key()) <
           0))) {
-      if (!level_files.empty()) {
+
+      if (level_iters_.size() >= size_t(level)) {
```
Comment (igorcanadi) [inline db/forward_iterator.cc]: Can you use new syntax for casting? `static_cast<size_t>(level)`

Comment (yuslepukhin) [inline db/forward_iterator.cc]: @igorcanadi will do

**Comment by rven1** [comment]
Can you assert that leveliters.size >= level
And verify that it deosn’t fail.
In RebuildIterator,  there is a reserve and the new number of levels will always be the same as what is reserved, since we don’t change the number of levels

**Comment by yuslepukhin** [comment]
@rven1 I can do that. However, that will fail probably. And the reason for that is reserve() is not the right function to use because it allocates memory but does not create elements to which you can assign. So the size will remain 0 the the acccess check will be out of bounds. resize() is a better choice which will init the vector with zeros.

**Comment by yuslepukhin** [comment]
@rven1 so if you find the problem there then my changes are not necessary and you could add asserts yourself.

**Comment by yuslepukhin** [comment]
@rven1 I just looked and yes, it is the problem. Change reserve() to resize() on line 499.

**Comment by siying** [comment]
@rven1 is fixing it: https://reviews.facebook.net/D50835

Our valgrind build was failing because of other issues so we didn't realize the regression.

**Pull request closed without merging**